File: src/domain/usecases/folders/CreateFolderUseCase.ts
import type { ImageFolderRepository } from '@/domain/repositories/ImageFolderRepository'
import type { AreaRepository } from '@/domain/repositories/AreaRepository'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import type { ImageFolder } from '@/domain/entities/ImageFolder'
import type { User } from '@/domain/entities/User'
import { UserRole } from '@/domain/value-objects/UserRole'
import { UnauthorizedError, ValidationError } from '@/domain/errors/DomainError'

export interface CreateFolderRequest {
  areaId: string
  name: string
  description: string
  assignToAllTechnicians: boolean
  assignedTechnicianIds: string[]
}

export class CreateFolderUseCase {
  private readonly folderRepository: ImageFolderRepository
  private readonly areaRepository: AreaRepository
  private readonly userRepository: UserRepository

  constructor(
    folderRepository: ImageFolderRepository,
    areaRepository: AreaRepository,
    userRepository: UserRepository,
  ) {
    this.folderRepository = folderRepository
    this.areaRepository = areaRepository
    this.userRepository = userRepository
  }

  async execute(actor: User, request: CreateFolderRequest): Promise<ImageFolder> {
    if (!actor.active) {
      throw new UnauthorizedError('Cuenta inactiva')
    }

    const name = request.name.trim()
    const description = request.description.trim()
    const areaId = request.areaId.trim()

    if (!areaId) {
      throw new ValidationError('Debes indicar el área')
    }
    if (!name) {
      throw new ValidationError('El nombre de la carpeta es obligatorio')
    }

    const area = await this.areaRepository.getById(areaId)
    if (!area) {
      throw new ValidationError('Área no encontrada')
    }

    const assignment = await resolveAssignments(
      this.userRepository,
      actor,
      request.assignToAllTechnicians,
      request.assignedTechnicianIds,
    )

    return this.folderRepository.create({
      areaId: area.id,
      areaName: area.name,
      name,
      description,
      ownerId: actor.id,
      ownerName: actor.displayName,
      assignToAllTechnicians: assignment.assignToAllTechnicians,
      assignedTechnicianIds: assignment.assignedTechnicianIds,
      assignedTechnicianNames: assignment.assignedTechnicianNames,
    })
  }
}

export async function resolveAssignments(
  userRepository: UserRepository,
  actor: User,
  assignToAllTechnicians: boolean,
  assignedTechnicianIds: string[],
): Promise<{
  assignToAllTechnicians: boolean
  assignedTechnicianIds: string[]
  assignedTechnicianNames: string[]
}> {
  if (assignToAllTechnicians) {
    return {
      assignToAllTechnicians: true,
      assignedTechnicianIds: [],
      assignedTechnicianNames: [],
    }
  }

  const uniqueIds = [...new Set(assignedTechnicianIds.map((id) => id.trim()).filter(Boolean))]

  // El creador técnico siempre queda incluido.
  if (actor.role === UserRole.Tecnico && !uniqueIds.includes(actor.id)) {
    uniqueIds.push(actor.id)
  }

  if (uniqueIds.length === 0) {
    throw new ValidationError(
      'Selecciona al menos un técnico o elige “Todos los técnicos”',
    )
  }

  const technicians = (await userRepository.listTechnicians()).filter(
    (user) => user.role === UserRole.Tecnico && user.active,
  )
  const byId = new Map(technicians.map((user) => [user.id, user]))

  const ids: string[] = []
  const names: string[] = []
  for (const id of uniqueIds) {
    const tech = byId.get(id)
    if (!tech) {
      throw new ValidationError('Hay un técnico inválido o inactivo en la asignación')
    }
    ids.push(tech.id)
    names.push(tech.displayName)
  }

  return {
    assignToAllTechnicians: false,
    assignedTechnicianIds: ids,
    assignedTechnicianNames: names,
  }
}

File: consorcio/lib/domain/usecases/create_folder_use_case.dart
import '../entities/app_user.dart';
import '../entities/image_folder.dart';
import '../errors/domain_exception.dart';
import '../repositories/area_repository.dart';
import '../repositories/image_folder_repository.dart';
import '../repositories/user_repository.dart';
import '../value_objects/geo_location.dart';
import '../value_objects/user_role.dart';

class CreateFolderUseCase {
  CreateFolderUseCase(
    this._folderRepository,
    this._areaRepository,
    this._userRepository,
  );
  final ImageFolderRepository _folderRepository;
  final AreaRepository _areaRepository;
  final UserRepository _userRepository;

  Future<List<AppUser>> listTechniciansForAssignment() {
    return _userRepository.listTechnicians();
  }

  Future<ImageFolder> execute(
    AppUser actor, {
    required String areaId,
    required String name,
    required String description,
    required GeoLocation location,
    required bool assignToAllTechnicians,
    required List<String> assignedTechnicianIds,
  }) async {
    actor.assertCanOperateApp();

    final cleanName = name.trim();
    if (cleanName.isEmpty) {
      throw DomainException('El nombre de la carpeta es obligatorio');
    }
    if (!location.isValid) {
      throw DomainException('La ubicación GPS no es válida');
    }

    final area = await _areaRepository.getById(areaId);
    if (area == null) {
      throw DomainException('Área no encontrada');
    }

    final assignment = await resolveFolderAssignments(
      userRepository: _userRepository,
      actor: actor,
      assignToAllTechnicians: assignToAllTechnicians,
      assignedTechnicianIds: assignedTechnicianIds,
    );

    return _folderRepository.create(
      CreateImageFolderInput(
        areaId: area.id,
        areaName: area.name,
        name: cleanName,
        description: description.trim(),
        ownerId: actor.id,
        ownerName: actor.displayName,
        assignToAllTechnicians: assignment.assignToAll,
        assignedTechnicianIds: assignment.ids,
        assignedTechnicianNames: assignment.names,
        location: location,
      ),
    );
  }
}

class FolderAssignmentResult {
  const FolderAssignmentResult({
    required this.assignToAll,
    required this.ids,
    required this.names,
  });

  final bool assignToAll;
  final List<String> ids;
  final List<String> names;
}

Future<FolderAssignmentResult> resolveFolderAssignments({
  required UserRepository userRepository,
  required AppUser actor,
  required bool assignToAllTechnicians,
  required List<String> assignedTechnicianIds,
}) async {
  if (assignToAllTechnicians) {
    return const FolderAssignmentResult(
      assignToAll: true,
      ids: [],
      names: [],
    );
  }

  final uniqueIds = <String>{
    ...assignedTechnicianIds.map((id) => id.trim()).where((id) => id.isNotEmpty),
  };
  uniqueIds.add(actor.id);

  if (uniqueIds.isEmpty) {
    throw DomainException(
      'Selecciona al menos un técnico o elige “Todos los técnicos”',
    );
  }

  final technicians = await userRepository.listTechnicians();
  final byId = {
    for (final tech in technicians)
      if (tech.role == UserRole.tecnico && tech.active) tech.id: tech,
  };

  final ids = <String>[];
  final names = <String>[];
  for (final id in uniqueIds) {
    final tech = byId[id];
    if (tech == null) {
      throw DomainException('Hay un técnico inválido o inactivo en la asignación');
    }
    ids.add(tech.id);
    names.add(tech.displayName);
  }

  return FolderAssignmentResult(
    assignToAll: false,
    ids: ids,
    names: names,
  );
}
